wallet/wallet-controller/src/sync/mod.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{sync::Arc, time::Duration};
+
+use common::{chain::ChainConfig, primitives::BlockHeight};
+use node_comm::node_traits::NodeInterface;
+use tokio::sync::Mutex;
+use wallet::DefaultWallet;
+
+#[derive(thiserror::Error, Debug)]
+pub enum SyncError {
+    #[error("Unexpected RPC error: {0}")]
+    UnexpectedRpcError(String),
+    #[error("Unexpected wallet error: {0}")]
+    UnexpectedWalletError(wallet::WalletError),
+}
+
+/// Sync the wallet state (blocks) from the node.
+/// Returns true if the wallet state has changed and false otherwise.
+async fn sync_blocks<T: NodeInterface>(
+    chain_config: &ChainConfig,
+    rpc_client: &mut T,
+    wallet: Arc<Mutex<DefaultWallet>>,
+) -> Result<bool, SyncError> {
+    // TODO: Make it more efficient: download blocks concurrently and send fewer requests
+    let wallet_block_height = match wallet
+        .lock()
+        .await
+        .get_best_block_height()
+        .map_err(SyncError::UnexpectedWalletError)?
+    {
+        Some(height) => height,
+        None => {
+            wallet.lock().await.scan_genesis().map_err(SyncError::UnexpectedWalletError)?;
+            return Ok(true);
+        }
+    };
+
+    let node_block_height = rpc_client
+        .get_best_block_height()
+        .await
+        .map_err(|e| SyncError::UnexpectedRpcError(e.to_string()))?;
+
+    if node_block_height <= wallet_block_height {
+        return Ok(false);
+    }
+
+    if wallet_block_height > BlockHeight::zero() {
+        let node_block_id = match rpc_client
+            .get_block_id_at_height(wallet_block_height)
+            .await
+            .map_err(|e| SyncError::UnexpectedRpcError(e.to_string()))?
+        {
+            Some(id) => id,
+            None => {
+                return Err(SyncError::UnexpectedRpcError(format!(
+                    "Node block height is {node_block_height} but block at height {wallet_block_height} is None"
+                )));
+            }
+        };
+        let wallet_block_id = match wallet
```

Comment:
I see what you're doing down there... you're doing that manually. You shouldn't. 

First, again, you're depending on the height. That's wrong.

Second, there's a function for doing that in chainstate (finding the common ancestor). Perhaps you should add a chainstate interface function and call it, which takes two arbitrary block ids, and returns the common ancestor.

Reply:
I added `get_last_common_height` and use it to sync the wallet. Block ids are used to determine if the node state has changed.